src/core/coap/coap_secure.hpp

```diff
@@ -61,6 +61,16 @@ class CoapSecure : public CoapBase
      */
     typedef void (*ConnectedCallback)(bool aConnected, void *aContext);
 
+    /**
+     * Pointer is called once DTLS connection is established or closed.
+     *
+     * @param[in]  aConnected  TRUE if a connection was established, FALSE otherwise.
+     * @param[in]  aWithError  TRUE if the connection was closed due to an error.
+     * @param[in]  aContext    A pointer to arbitrary context information.
+     *
+     */
+    typedef void (*ExtendedConnectedCallback)(bool aConnected, bool aWithError, void *aContext);
```

Comment:
I'm not familiar with the concepts in CoAP. Could you elaborate on why it's called 'extended'? 

Is it called 'extended' because it's has more capability than the `ConnectedCallback`? If so, probably we can just refactor `ConnectedCallback` instead of introducing a new API?

Reply:
+1, I agree. Better to update `ConnectCallback`.

Reply:
I added an extended callback to `ConnectCallback` is because the API `otCoapSecureSetClientConnectedCallback` uses `otHandleCoapSecureClientConnect` to set the ConnectedCallback. If I update the ConnectedCallback I need to update the `otHandleCoapSecureClientConnect` API. In the openthread project I don't find callers of `otCoapSecureSetClientConnectedCallback` though, but I wonder if updating them would break use cases in other projects using openthread.

What's the general guideline for updating APIs?

Reply:
I updated the `ConnectCallback` together with the `otHandleCoapSecureClientConnect` APIs given that this change should be easy for callers to fix.